File: Ordination/C++/GnomeSort.cpp
#include <iostream>


template <typename T> class GnomeSort{
	public:
		void sort(T array[],int leftIndex,int rightIndex,int size){
			if(validateParams(array,leftIndex,rightIndex,size)){
				//this->simpleGnomeSort(array,leftIndex,rightIndex);
				this->recursiveGnomeSort(array,leftIndex,rightIndex,leftIndex + 1);
				this->showArray(array,size);
			}else{
				std::cout << "Nao era para entrar aqui!!" <<std::endl;
			}
		}
		void sort(T array[],int size){
			this->sort(array,0,size -1,size);
		}
	private:
		
		void simpleGnomeSort(T array[],int leftIndex,int rightIndex){
			int pivot = leftIndex + 1;
			while(pivot <= rightIndex){
				if(array[pivot] >= array[pivot-1]){
					pivot++;
				}else{
					swap(array,pivot,pivot -1);
					if(pivot > leftIndex + 1){
						pivot--;	
					}else{
						pivot++;
					}
				}
			}
		}
	
		void recursiveGnomeSort(T array[],int leftIndex,int rightIndex,int pivot){
			if(pivot <= rightIndex){
				if(array[pivot] >= array[pivot - 1]){
					this->recursiveGnomeSort(array,leftIndex,rightIndex,pivot + 1);
				}else{
					this->swap(array,pivot,pivot - 1);
					if(pivot > leftIndex +1){
						this->recursiveGnomeSort(array,leftIndex,rightIndex,pivot- 1);
					}else{
						this->recursiveGnomeSort(array,leftIndex,rightIndex,pivot + 1);
					}
				}
			}
		}

		void swap(T array[],int i,int j){
			T aux = array[i];
			array[i] = array[j];
			array[j] = aux;
		}
		bool validateParams(T array[],int leftIndex,int rightIndex,int size){
			if(array == NULL || leftIndex < 0 || rightIndex > size - 1 || leftIndex >= rightIndex){
				return false;
			}else{
				return true;
			}
		}
		void showArray(T array[],int size){
			std::cout << "Array: [";
			for(int i = 0 ; i < size;i++){
				if(i != size - 1){
					std::cout << array[i] << ",";
				}else{
					std::cout << array[i] << "]";
				}
			}
			std::cout << std::endl;
		}
};

File: Java/src/GnomeSort/GnomeSort.java
package GnomeSort;

import sorting.AbstractSort;

public class GnomeSort<T extends Comparable<T>> extends AbstractSort<T> {

	@Override
	public void sort(T[] array, int leftIndex, int rightIndex) {
		if(validateParams(array, leftIndex, rightIndex)){
			//simpleGnomeSort(array, leftIndex, rightIndex);
			recursiveGnomeSort(array, leftIndex, leftIndex +1, rightIndex);
		}else{
			//Do nothing
		}
	}
	private void simpleGnomeSort(T[] array,int leftIndex,int rightIndex){
		int pivot = leftIndex +1;
		while(pivot <= rightIndex){
			if(array[pivot].compareTo(array[pivot - 1]) >= 0){
				pivot++;
			}else{
				swap(array,pivot,pivot-1);
				if(pivot > leftIndex +1){
					pivot--;
				}else{
					pivot++;
				}
			}
		}
	}
	
	private void recursiveGnomeSort(T[] array,int leftIndex,int pivot,int rightIndex){
		if(pivot <= rightIndex){
			if(array[pivot].compareTo(array[pivot -1]) >= 0){
				recursiveGnomeSort(array, leftIndex, pivot +1, rightIndex);
			}else{
				swap(array,pivot,pivot -1);
				if(pivot > leftIndex +1){
					recursiveGnomeSort(array, leftIndex, pivot-1, rightIndex);
				}else{
					recursiveGnomeSort(array, leftIndex, pivot +1, rightIndex);
				}
			}
		}
	}
	
	private void swap(T[] array,int i , int j){
		T aux = array[i];
		array[i] = array[j];
		array[j] = aux;
	}
	
	private boolean validateParams(T[] array,int leftIndex,int rightIndex){
		if(array == null || leftIndex < 0 || rightIndex > array.length - 1 || leftIndex >= rightIndex){
			return false;
		}else{
			return true;
		}
	}
}
